### PR TITLE
[IMPROVEMENT] Pipefy - Create Card requires user to input GraphQL query

### LINKED ISSUE #4194: [IMPROVEMENT] Pipefy - Create Card requires user to input GraphQL query
**Is your feature request related to a problem? Please describe.**
The Pipefy - Create Card requires user to input GraphQL query, which is unfriendly, and the user doesn't know how to input the query
image

**Describe the solution you'd like**
We should allow user to input Pipe, then we fetch the Card structure in the Pipe and renders the props respectively. For example:
- The User selects Pipe: Pipedream Test
- The action renders the additional props matching with Card structure in the Pipe such as: Title, parent, phase, etc...

**Do you have a workaround?**
Currently user only able to input GraphQL query

**Comparable features in other tools?**
Zapier allow the user to select Pipe
image
Then input values per the card structure
FireShot Capture 245 - Add cards to Pipefy for new rows on Google Sheets - Zapier_ - workflows zapier com

### PR DESCRIPTION
**Pull request by michelle0927**
Resolves #4194 

Updates the Pipefy Create Card action to list the available fields as additionalProps instead of requiring the user to enter a GraphQL query.

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4311"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/pipefy/pipefy.app.mjs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        "User-Agent": "@PipedreamHQ/pipedream v0.1",
+      };
+    },
+    async _makeRequest(data, endpoint, $ = this) {
+      const config = {
+        method: "POST",
+        url: `${this._getBaseUrl()}/${endpoint}`,
+        headers: this._getHeaders(),
+        data,
+      }; console.log(config);
```
Comment (andrewjschuang) [inline components/pipefy/pipefy.app.mjs:61]:
```suggestion
      };
```

File: components/pipefy/pipefy.app.mjs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+      return axios($, config);
+    },
+    async _makeQueriesRequest(data = null) {
+      return this._makeRequest(data, "queries");
+    },
+    async _makeGraphQlRequest(data = null) {
```
Comment (andrewjschuang) [inline components/pipefy/pipefy.app.mjs:67]: No need for this initialization

```suggestion
    async _makeGraphQlRequest(data) {
```

File: components/pipefy/pipefy.app.mjs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    async _makeRequest(data, endpoint, $ = this) {
+      const config = {
+        method: "POST",
+        url: `${this._getBaseUrl()}/${endpoint}`,
+        headers: this._getHeaders(),
+        data,
+      }; console.log(config);
+      return axios($, config);
+    },
+    async _makeQueriesRequest(data = null) {
```
Comment (andrewjschuang) [inline components/pipefy/pipefy.app.mjs:64]: No need for this initialization

```suggestion
    async _makeQueriesRequest(data) {
```

File: components/pipefy/actions/create-card/create-card.mjs [hunk 4]
```diff
@@ -35,17 +74,34 @@ export default {
...
 
-    return await axios($, {
-      method: "post",
-      url: "https://api.pipefy.com/graphql",
-      headers: {
-        Authorization: `Bearer ${this.pipefy.$auth.token}`,
-      },
-      data: this.graphql_query,
-    });
+    return await this.pipefy._makeRequest(data, "graphql");
```
Comment (andrewjschuang) [inline components/pipefy/actions/create-card/create-card.mjs:105]:
```suggestion
    return await this.pipefy._makeRequest(data, "graphql", $);
```

Actually, creating a method and calling it here would be better

File: components/pipefy/actions/create-card/create-card.mjs [hunk 5]
```diff
@@ -35,17 +74,34 @@ export default {
   }
   */
 
-    if (!this.graphql_query) {
-      throw new Error("Must provide graphql_query parameter.");
+    let mutationString = `
```
Comment (andrewjschuang) [inline components/pipefy/actions/create-card/create-card.mjs:77]: Is there a better way to create a query with variables?
Maybe like in **Shopify Partner** or **Shopify**?

**Review by andrewjschuang — CHANGES_REQUESTED** [review]
Hi @michelle0927, looking very good, I've added just a few comments

**Review by andrewjschuang — DISMISSED** [review]
Thanks, looks great to me!

**Review by vellames — DISMISSED** [review]
lgtm

**Comment by vunguyenhung** [comment]
Hello @michelle0927, I have tested this PR and there's only 1 test cases failed. Then it's ready for release.

Please check the test report below for more information
[Pipefy_4311_2202.pdf](https://github.com/PipedreamHQ/pipedream/files/9675861/Pipefy_4311_2202.pdf)

Great job!

**Comment by michelle0927** [comment]
@vunguyenhung I created a Pipe using the Recruitment Process template, and confirmed that the "recruiter" field is expecting an `id` instead of a `string`. If you enter the `id` of a member/user, you shouldn't see that error. I added the field `type` to the prop descriptions to give the user a better idea of what to enter.

I also discovered that the pipe's "start_fields" are required to create a card, so I added those fields in additionalProps as well. (Hint: Candidate expects an id as well).

**Comment by vunguyenhung** [comment]
Hello @michelle0927, thank you for explaining! I'm have test the Create Card action again following your explanation and it passed! Ready for release!

New Test Report
[Pipefy_4311_2031.pdf](https://github.com/PipedreamHQ/pipedream/files/9684572/Pipefy_4311_2031.pdf)

**Pull request merged**